pkg/minikube/bootstrapper/kubeadm/dns/coredns.go

```diff
@@ -0,0 +1,346 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// minikube-specific CoreDNS manifests based on the default kubeadm's embedded ones
+// ref: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go
+
+package dns
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"html/template"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"k8s.io/klog/v2"
+	"k8s.io/minikube/pkg/kapi"
+	"k8s.io/minikube/pkg/minikube/assets"
+	"k8s.io/minikube/pkg/minikube/bootstrapper/images"
+	"k8s.io/minikube/pkg/minikube/command"
+	"k8s.io/minikube/pkg/minikube/config"
+	"k8s.io/minikube/pkg/minikube/vmpath"
+	kconst "k8s.io/minikube/third_party/kubeadm/app/constants"
+)
+
+const (
+	// CoreDNSService is the CoreDNS Service manifest
+	CoreDNSService = `
```

Comment:
@prezha one problem I feel like might happen, how do we generate this yaml file and with future kubernetes versions is it gonna change this yaml and we will need to remember to regenrate it ? is there an automated way of generating this to future proof it ?

Reply:
and also if user chooses an older kubernetes we wanna make sure this yaml works there too or we could use the Old way for Non-Default Kubernets version (the latest that runs by default)

Reply:
@medyagh i thought about and accounted for that:

based on the [history of changes](https://github.com/kubernetes/kubernetes/commits/master/cmd/kubeadm/app/phases/addons/dns/manifests.go), in the past 2+ years, there were no frequent or significant changes to the kubeadm's embedded coredns deployment manifest, which is well over our target extended support policy of last 6 minor versions
based on this, we do not have an automated way to update coredns manifest atm, but i intentionally made it in diff-friendly format so it would be straight-forward finding any differences and upgrade if/when needed in the future - namely:
- minikube's coredns: https://github.com/kubernetes/minikube/blob/8de81377cf7b0f8f4f5586fc099dc6812fe2ec91/pkg/minikube/bootstrapper/kubeadm/dns/coredns.go
- kubeadm's coredns: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/dns/manifests.go

this would be the default behaviour, and should a user want/need to use the old way - they can do so just by using the `--disable-optimizations` flag

last but not least, as a potential next step, with this "separation", we could relatively easily enable users to provide their own coredns manifest (we'd still need eg, to patch it to inject ip address of host.minikube.internal that's only known during the runtime)